File: Backend/Microservices/Bursary/Eefa.Bursary.Application/UseCases/Definitions/BankAccount/Commands/Add/CreateBankAccountCommand.cs
using Eefa.Common;
using Eefa.Common.CommandQuery;
using MediatR;
using Eefa.Bursary.Domain.Entities.Definitions;
using System.Threading.Tasks;
using System.Threading;
using AutoMapper;
using Eefa.Bursary.Infrastructure.Interfaces;
using Eefa.Common.Exceptions;
using System;
using Eefa.Bursary.Application.UseCases.Definitions.BankBranch.Commands.Add;

namespace Eefa.Bursary.Application.UseCases.Definitions.BankAccount.Commands.Add
{
    public class CreateBankAccountCommand : CommandBase, IRequest<ServiceResult<BankAccounts>>, IMapFrom<CreateBankAccountCommand>, ICommand
    {
        public int? ParentId { get; set; }
        public int BankBranchId { get; set; } = default!;
        public string Sheba { get; set; } = default!;
        public string AccountNumber { get; set; } = default!;
        public int? SubsidiaryCodeId { get; set; }
        public int? RelatedBankAccountId { get; set; }
        public int AccountTypeBaseId { get; set; } = default!;
        public int AccountHeadId { get; set; }
        public int? AccountReferencesGroupId { get; set; }
        public int? ReferenceId { get; set; }
        public int? AccountStatus { get; set; }
        public decimal WithdrawalLimit { get; set; } = 0;
        public bool HaveChekBook { get; set; } = default!;
        public int CurrenceTypeBaseId { get; set; } = default!;
        public string SignersJson { get; set; } = "";

        public void Mapping(Profile profile)
        {
            profile.CreateMap<CreateBankAccountCommand, BankAccounts>().IgnoreAllNonExisting();
        }

    }

    public class CreateBankAccountCommandHandler : IRequestHandler<CreateBankAccountCommand, ServiceResult<BankAccounts>>
    {
        private readonly IMapper _mapper;
        private readonly IBursaryUnitOfWork _uow;

        public CreateBankAccountCommandHandler(IMapper mapper, IBursaryUnitOfWork uow)
        {
            _mapper = mapper;
            _uow = uow;
        }

        public async Task<ServiceResult<BankAccounts>> Handle(CreateBankAccountCommand request, CancellationToken cancellationToken)
        {
            var bankaccount = _mapper.Map<BankAccounts>(request);

            if (bankaccount == null)
            {
                throw new ValidationError("اطلاعات ورودی ارسال نگردیده است");
            }
            _uow.BankAccounts.Add(bankaccount);
            var value = await _uow.SaveChangesAsync(cancellationToken);

            if (value <= 0)
                throw new Exception("بروز خطا در ثبت اطلاعات بانک");

            return ServiceResult<BankAccounts>.Success(bankaccount);
        }
    }


}

File: Frontend/angular-app/src/app/modules/bursary/repositories/bank-account/commands/create-bank-account-command.ts
import {Inject} from "@angular/core";
import {ServiceResult} from "../../../../../core/models/service-result";
import {IRequest, IRequestHandler} from "../../../../../core/services/mediator/interfaces";
import {HttpService} from "../../../../../core/services/http/http.service";
import {ApplicationError} from "../../../../../core/exceptions/application-error";
import {MediatorHandler} from "../../../../../core/services/mediator/decorator";
import {ValidationRule} from "../../../../../core/validation/validation-rule";
import {HttpRequest} from "../../../../../core/services/http/http-request";
import {NotificationService} from "../../../../../shared/services/notification/notification.service";
import {BankAccount} from "../../../entities/bank-account";

export class CreateBankAccountCommand extends IRequest<CreateBankAccountCommand, BankAccount> {
  parentId: number | undefined = undefined;
  bankBranchId: number | undefined = undefined;
  bankBranchTitle: string | undefined = undefined;
  title: string | undefined = undefined;
  sheba: string | undefined = undefined;
  accountNumber: string | undefined = undefined;
  subsidiaryCodeId: number | undefined = undefined;
  relatedBankAccountId: number | undefined = undefined;
  accountTypeBaseId: number | undefined = undefined;
  accountReferenceGroupId: number | undefined = undefined;
  accountStatus: number | undefined = undefined;
  withdrawalLimit: number | undefined = undefined;
  haveChekBook: boolean | undefined = undefined;
  currenceTypeBaseId: number | undefined = undefined;
  signersJson: string | undefined = undefined;


  constructor() {
    super();
  }

  mapFrom(entity: BankAccount): CreateBankAccountCommand {
    throw new ApplicationError(CreateBankAccountCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  mapTo(): BankAccount {
    throw new ApplicationError(CreateBankAccountCommand.name, this.mapTo.name, 'Not Implemented Yet')
  }

  get url(): string {
    return "/bursary/bankAccounts/add";
  }

  get validationRules(): ValidationRule[] {
    return [];
  }
}

@MediatorHandler(CreateBankAccountCommandHandler.name)
export class CreateBankAccountCommandHandler implements IRequestHandler<CreateBankAccountCommand, BankAccount> {
  constructor(
    @Inject(HttpService) private _httpService: HttpService,
    @Inject(NotificationService) private _notificationService: NotificationService
  ) {
  }

  async Handle(request: CreateBankAccountCommand): Promise<BankAccount> {
    let httpRequest: HttpRequest<CreateBankAccountCommand> = new HttpRequest<CreateBankAccountCommand>(request.url, request);
    // httpRequest.Query  += ;
    // httpRequest.Headers = httpRequest.Headers.append()

    return await this._httpService.Post<CreateBankAccountCommand, ServiceResult<BankAccount>>(httpRequest).toPromise().then(response => {
      this._notificationService.showSuccessMessage()
      return response.objResult
    })
  }
}
